common/deadlock/deadlock.go

```diff
@@ -0,0 +1,222 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deadlock
+
+import (
+	"context"
+	"runtime/pprof"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/fx"
+	"google.golang.org/grpc/health"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/internal/goro"
+	"go.temporal.io/server/service/history/shard"
+)
+
+type (
+	params struct {
+		fx.In
+
+		Logger       log.SnTaggedLogger
+		Collection   *dynamicconfig.Collection
+		HealthServer *health.Server
+
+		// root pingables:
+		NamespaceRegistry namespace.Registry
+		ClusterMetadata   cluster.Metadata
+		ShardController   shard.Controller `optional:"true"`
+	}
+
+	config struct {
+		DumpGoroutines    dynamicconfig.BoolPropertyFn
+		FailHealthCheck   dynamicconfig.BoolPropertyFn
+		AbortProcess      dynamicconfig.BoolPropertyFn
+		Interval          dynamicconfig.DurationPropertyFn
+		MaxWorkersPerRoot dynamicconfig.IntPropertyFn
+	}
+
+	deadlockDetector struct {
+		logger       log.Logger
+		healthServer *health.Server
+		config       config
+		roots        []common.Pingable
+		loops        goro.Group
+	}
+
+	loopContext struct {
+		dd      *deadlockDetector
+		root    common.Pingable
+		ch      chan common.PingCheck
+		workers int32
+		ctx     context.Context // cancelled when loop should exit
+	}
+)
+
+func NewDeadlockDetector(params params) *deadlockDetector {
+	roots := []common.Pingable{
+		params.NamespaceRegistry,
+		params.ClusterMetadata,
+	}
+	if params.ShardController != nil {
+		roots = append(roots, params.ShardController)
+	}
+	return &deadlockDetector{
+		logger:       params.Logger,
+		healthServer: params.HealthServer,
+		config: config{
+			DumpGoroutines:    params.Collection.GetBoolProperty(dynamicconfig.DeadlockDumpGoroutines, true),
+			FailHealthCheck:   params.Collection.GetBoolProperty(dynamicconfig.DeadlockFailHealthCheck, true),
+			AbortProcess:      params.Collection.GetBoolProperty(dynamicconfig.DeadlockAbortProcess, false),
+			Interval:          params.Collection.GetDurationProperty(dynamicconfig.DeadlockInterval, 30*time.Second),
+			MaxWorkersPerRoot: params.Collection.GetIntProperty(dynamicconfig.DeadlockMaxWorkersPerRoot, 10),
+		},
+		roots: roots,
+	}
+}
+
+func (dd *deadlockDetector) Start() error {
+	for _, root := range dd.roots {
+		loopCtx := &loopContext{
+			dd:   dd,
+			root: root,
+			ch:   make(chan common.PingCheck),
+		}
+		dd.loops.Go(loopCtx.run)
+	}
+	return nil
+}
+
+func (dd *deadlockDetector) Stop() error {
+	dd.loops.Cancel()
+	// don't wait for workers to exit, they may be blocked
+	return nil
+}
+
+func (dd *deadlockDetector) detected(name string) {
+	dd.logger.Error("deadlock detected", tag.Name(name))
```

Comment:
nit: maybe "potential deadlock detected" as we don't know for sure this is deadlock, it is just timeout